kv/kvclient/rangefeed: TestRangefeedCatchupStarvation failed

kv/kvclient/rangefeed.TestRangefeedCatchupStarvation [failed](https://mesolite.cluster.engflow.com/invocations/default/107ae9ca-88eb-4586-8168-c21df1671ebc?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZjbGllbnQvcmFuZ2VmZWVkOnJhbmdlZmVlZF90ZXN0) on master @ [b13396d63f638bd109dad208337dd7a443a921e0](https://github.com/cockroachdb/cockroach/commits/b13396d63f638bd109dad208337dd7a443a921e0):


```
=== RUN   TestRangefeedCatchupStarvation
    test_log_scope.go:172: test logs captured to: outputs.zip/logTestRangefeedCatchupStarvation1559085369
    test_log_scope.go:82: use -show-logs to present logs inline
    rangefeed_external_test.go:1988: -- test log scope end --
test logs left over in: outputs.zip/logTestRangefeedCatchupStarvation1559085369
W260529 12:06:54.876484 138326 kv/kvclient/rangefeed/scanner.go:111  [rangefeed=foo] 1  failed to perform initial scan: boom
W260529 12:06:54.880842 138338 kv/kvclient/rangefeed/rangefeed.go:408  [rangefeed=foo] 2  rangefeed failed 0 times, restarting: boom
--- FAIL: TestRangefeedCatchupStarvation (375.57s)
=== RUN   TestRangefeedCatchupStarvation/feed_type=mux/buffered_sender=false
    test_server_shim.go:168: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    test_server_shim.go:637: DRPC is enabled (via testserver factory defaults)
    rangefeed_external_test.go:1984: test timed out
--- FAIL: TestRangefeedCatchupStarvation/feed_type=mux/buffered_sender=false (296.20s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRangefeedCatchupStarvation.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-64353

--- a/pkg/kv/kvclient/rangefeed/BUILD.bazel
+++ b/pkg/kv/kvclient/rangefeed/BUILD.bazel
@@ -89,6 +89,7 @@ go_test(
         "//pkg/testutils/sqlutils",
         "//pkg/testutils/storageutils",
         "//pkg/testutils/testcluster",
+        "//pkg/util",
         "//pkg/util/ctxgroup",
         "//pkg/util/encoding",
         "//pkg/util/hlc",

--- a/pkg/kv/kvclient/rangefeed/rangefeed_external_test.go
+++ b/pkg/kv/kvclient/rangefeed/rangefeed_external_test.go
@@ -38,6 +38,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/storageutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
+	"github.com/cockroachdb/cockroach/pkg/util"
 	"github.com/cockroachdb/cockroach/pkg/util/encoding"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
@@ -1899,9 +1900,21 @@ func TestRangefeedCatchupStarvation(t *testing.T) {
 		settings := cluster.MakeTestingClusterSettings()
 		kvserver.RangefeedUseBufferedSender.Override(ctx, &settings.SV, rt.useBufferedSender)
 		kvserver.RangefeedEnabled.Override(ctx, &settings.SV, true)
-		// Lower the limit to make it more likely to get starved.
+		// Lower the limits to make it more likely to get starved. The
+		// per-consumer limit must stay strictly below the number of
+		// catch-up slots a single priority band can occupy, otherwise one
+		// consumer can hold every slot and the second consumer is starved.
+		//
+		// None of the rangefeeds below set WithSystemTablePriority, so they
+		// all run at BulkNormalPri and share the catch-up queue's
+		// low-priority band. With catch-up prioritization enabled (the
+		// default), that band is capped at catchupQueueLowMaxActive(8) = 6
+		// of the 8 total iterator slots; the other 2 are reserved for
+		// high-priority (system-table) rangefeeds. A per-consumer limit of 4
+		// therefore leaves consumer-2 at least 6-4=2 slots regardless of how
+		// many catch-up scans consumer-1 floods the queue with.
 		kvserver.ConcurrentRangefeedItersLimit.Override(ctx, &settings.SV, 8)
-		kvserver.PerConsumerCatchupLimit.Override(ctx, &settings.SV, 6)
+		kvserver.PerConsumerCatchupLimit.Override(ctx, &settings.SV, 4)
 		srv, _, db := serverutils.StartServer(t, base.TestServerArgs{
 			Settings: settings,
 		})
@@ -1913,8 +1926,18 @@ func TestRangefeedCatchupStarvation(t *testing.T) {
 		mkKey := func(k string) roachpb.Key {
 			return encoding.EncodeStringAscending(scratchKey, k)
 		}
+		// The number of ranges drives the contention this test exercises: each
+		// range registers its own catch-up scan, so 32 ranges across consumer-1's
+		// many rangefeeds is what saturates the iterator pool. keysPerRange only
+		// drives delivery volume (consumer-2 must receive every key over a
+		// blocking channel), which dominates wall-clock time. Under race that
+		// volume pushes the test past its timeout when stressed, so trim it there
+		// while keeping the range count, and thus the contention, intact.
 		ranges := 32
 		keysPerRange := 128
+		if util.RaceEnabled {
+			keysPerRange = 24
+		}
 		totalKeys := ranges * keysPerRange
 		for i := range ranges {
 			for j := range keysPerRange {

